Add OpenZIM support

dnohales opened the pull request:
This is a WIP for the ZIM support in `libdmodel`.

- [x] Refactor code to support multiple types of shard (multiple implementation of `DmShard` interface)
- [x] Add `DmShardOpenZim` for a `zim-glib` based implementation
- [x] Figure out how to deal with the compatibility break for exposing `DmDomain->shards` that now is no longer a list of `EosShardShardFile` but a list of `DmShard`.
- [x] Implement `dm_domain_test_link` and `dm_domain_setup_link_tables` (adding the corresponding code in `DmShard` and its implementation).
- [x] Finish to adapt the `eknvfs` code.
- [x] Research changes needed in `EknServicesMultiplexer`.
- [x] Update/add tests.

https://phabricator.endlessm.com/T26276

dnohales committed:
Make abstraction for support multiple types of shards

dylanmccall commented on dmodel/dm-utils.c:
```diff
@@ -268,25 +283,35 @@ dm_utils_free_gparam_array (GArray *params)
+  if (g_strcmp0 (scheme, "ekn") == 0)
+   {
+     // Expecting ekn://[domain]/<object ID>[/member name]
+     g_auto(GStrv) tokens = g_strsplit (uri + strlen("ekn://"), "/", -1);
```
In line with the code below for zim URIs, it would probably be a smidge safer here to set a max_tokens of 3.

dylanmccall commented on dmodel/dm-utils.c:
```diff
@@ -268,25 +283,35 @@ dm_utils_free_gparam_array (GArray *params)
+   {
+     // Expecting ekn://[domain]/<object ID>[/member name]
+     g_auto(GStrv) tokens = g_strsplit (uri + strlen("ekn://"), "/", -1);
+     return g_strdup (tokens[1]);
```
The comment describing the URI format doesn't line up with the code here. If we're expecting an optional domain, we need to handle a few different cases based on g_strv_length (and, I suppose, whether the first token matches a domain). Looking at https://github.com/endlessm/libdmodel/blob/master/eknvfs/ekn-vfs.c#L257, I think we kind of ignore the optional domain in eknvfs, as well, so not sure if it's necessary to support it, in which case we could probably just update that comment and add it to the release notes. (Any ideas how we can find that out?). But either way I feel unsure about blindly returning tokens[1] since it could be undefined.

dylanmccall replied:
I was happy with the URI parsing implementation I came up with in this pull request. It spends quite a few CPU cycles being extra defensive, but probably not too much in the grand scheme of things:

https://github.com/endlessm/libdmodel/pull/12/files

dnohales replied:
Right now I'm gonna play with the eknvfs code and try to revisit this. Probably even try to reuse the functions if I can.

dylanmccall commented on meson.build:
```diff
@@ -36,6 +36,7 @@ libendless = dependency('endless-0', version: libendless_req)
 shard = dependency('eos-shard-0', version: shard_req)
 soup = dependency('libsoup-2.4')
 xapian_glib = dependency('xapian-glib-2.0')
+zim_glib = dependency('zim-glib-3.0')
```
Extremely pedantic detail: these dependencies aren't required for the code I see in this PR (yet). But if you're planning to add the Zim stuff in the same branch, carry on! :)

dnohales replied:
My bad! this should be moved to a different commit.

dylanmccall reviewed (commented):
DmShardEosShard makes me dizzy, but these look like good changes so far :) Using URIs more consistently seems like a good way to go. Were there any parts in particular you could use a second set of eyes on? I didn't get a chance to play with this outside of reading the code, but I'll be happy to do that on Monday.

dnohales commented:
> DmShardEosShard makes me dizzy

It's horrible :D

I'm all for renaming this if we find a better name. `DmCollection` maybe?

dnohales committed:
Ignore deprecations of GParameter

dnohales committed:
Allow to set most of the DmQuery properties

dnohales committed:
Add OpenZIM support

dnohales committed:
Ignore most common build directory

dylanmccall commented:
@dnohales can I merge this, or is there some stuff you'd still like to do?

dnohales commented:
Well... I was waiting for a review and tests are missing, but I think it could be merged if this is blocking the release. No apps are using SDK-6 for now. We could start porting those slowly to catch the bugs.

dnohales committed:
tests: Remove tap.py

There are absolutely no benefits of using tap.py. The output is harder
to read, no colored and when you print something in libdmodel, the tests
crash.

I suppose I'm removing this after being angry on spending a lot of time
on seeing why my test was failing.

dnohales committed:
Don't fail if xapian_databases.[].offset is missing

dnohales committed:
tests: Add tests for DmShardOpenZim

dnohales commented:
Added tests, ready to merge.

dnohales commented:
I'm not sure why the new Jenkins job is failing though.
